gcc-4.7.3 powerpc uclibc math library patch

Copyright (C) 2014 Curt Brune <curt@cumulusnetworks.com>

SPDX-License-Identifier:     GPL-2.0

On powerpc libgcc_s references copysignl which is a libm function but
glibc apparently also provides it via libc.  uClibc, however, puts it
in libm.

Add a configure test that adds -lm to the linker libraries for
powerpc-*-uclibc targets.

diff --git a/patches/gcc/4.7.3/100-gcc-4.7.3-powerpc-uclibc-math-library.patch b/patches/gcc/4.7.3/100-gcc-4.7.3-powerpc-uclibc-math-library.patch
new file mode 100644
index 0000000..deb4b9c
--- /dev/null
+++ b/patches/gcc/4.7.3/100-gcc-4.7.3-powerpc-uclibc-math-library.patch
@@ -0,0 +1,110 @@
+gcc-4.7.3 powerpc uclibc math library patch
+
+From: Curt Brune <curt@cumulusnetworks.com>
+
+On powerpc libgcc_s references copysignl which is a libm function but
+glibc apparently also provides it via libc.  uClibc, however, puts it
+in libm.
+
+Add a configure test that adds -lm to the linker libraries for
+powerpc-*-uclibc targets.
+---
+ libgcc/Makefile.in      |    4 +++-
+ libgcc/config/t-slibgcc |    2 +-
+ libgcc/configure        |   39 +++++++++++++++++++++++++++++++++++++++
+ 3 files changed, 43 insertions(+), 2 deletions(-)
+
+diff --git a/libgcc/Makefile.in b/libgcc/Makefile.in
+index e1cac49..63a8f98 100644
+--- a/libgcc/Makefile.in
++++ b/libgcc/Makefile.in
+@@ -41,6 +41,7 @@ long_double_type_size = @long_double_type_size@
+ decimal_float = @decimal_float@
+ enable_decimal_float = @enable_decimal_float@
+ fixed_point = @fixed_point@
++LIBGCC_LIBM = @LIBGCC_LIBM@
+ 
+ host_noncanonical = @host_noncanonical@
+ target_noncanonical = @target_noncanonical@
+@@ -928,9 +929,10 @@ libgcc_s$(SHLIB_EXT): $(libgcc-s-objects) $(extra-parts) libgcc.a
+ 		@multilib_dir@,$(MULTIDIR),$(subst \
+ 		@shlib_objs@,$(objects) libgcc.a,$(subst \
+ 		@shlib_base_name@,libgcc_s,$(subst \
++		@libgcc_libm@,$(LIBGCC_LIBM),$(subst \
+ 		@shlib_map_file@,$(mapfile),$(subst \
+ 		@shlib_slibdir_qual@,$(MULTIOSSUBDIR),$(subst \
+-		@shlib_slibdir@,$(shlib_slibdir),$(SHLIB_LINK))))))))
++		@shlib_slibdir@,$(shlib_slibdir),$(SHLIB_LINK)))))))))
+ 
+ libunwind$(SHLIB_EXT): $(libunwind-s-objects) $(extra-parts)
+ 	# @multilib_flags@ is still needed because this may use
+diff --git a/libgcc/config/t-slibgcc b/libgcc/config/t-slibgcc
+index 3727244..aa7f265 100644
+--- a/libgcc/config/t-slibgcc
++++ b/libgcc/config/t-slibgcc
+@@ -27,7 +27,7 @@ SHLIB_MAP = @shlib_map_file@
+ SHLIB_OBJS = @shlib_objs@
+ SHLIB_DIR = @multilib_dir@
+ SHLIB_SLIBDIR_QUAL = @shlib_slibdir_qual@
+-SHLIB_LC = -lc
++SHLIB_LC = @libgcc_libm@ -lc
+ SHLIB_MAKE_SOLINK = $(LN_S) $(SHLIB_SONAME) $(SHLIB_DIR)/$(SHLIB_SOLINK)
+ SHLIB_INSTALL_SOLINK = $(LN_S) $(SHLIB_SONAME) \
+ 	$(DESTDIR)$(slibdir)$(SHLIB_SLIBDIR_QUAL)/$(SHLIB_SOLINK)
+diff --git a/libgcc/configure b/libgcc/configure
+index d01dbbf..c62e084 100644
+--- a/libgcc/configure
++++ b/libgcc/configure
+@@ -555,6 +555,7 @@ ac_unique_file="static-object.mk"
+ ac_includes_default='/* none */'
+ ac_subst_vars='LTLIBOBJS
+ LIBOBJS
++LIBGCC_LIBM
+ asm_hidden_op
+ extra_parts
+ cpu_type
+@@ -4445,6 +4446,44 @@ if test -z "${sfp_machine_header}"; then
+ 	fi
+ fi
+ 
++# On powerpc libgcc_s references copysignl which is a libm function but
++# glibc apparently also provides it via libc as opposed to uClibc where
++# it lives in libm.
++{ $as_echo "$as_me:${as_lineno-$LINENO}: checking for library containing copysignl" >&5
++$as_echo_n "checking for library containing copysignl... " >&6; }
++if test "${libgcc_cv_copysignl_lib+set}" = set; then
++  $as_echo_n "(cached) " >&6
++else
++
++    echo '#include <features.h>' > conftest.c
++    case $target_alias in
++    *powerpc*)
++	echo 'int the_libc = __UCLIBC__;' >> conftest.c
++	;;
++    *)
++	echo 'int the_libc; /* dummy */' >> conftest.c
++	;;
++    esac
++    libgcc_cv_copysignl_lib="-lc"
++    if { ac_try='${CC-cc} -S conftest.c -o conftest.s 1>&5'
++  { { eval echo "\"\$as_me\":${as_lineno-$LINENO}: \"$ac_try\""; } >&5
++  (eval $ac_try) 2>&5
++  ac_status=$?
++  $as_echo "$as_me:${as_lineno-$LINENO}: \$? = $ac_status" >&5
++  test $ac_status = 0; }; }; then
++	libgcc_cv_copysignl_lib="-lm"
++    fi
++    rm -f conftest.*
++
++fi
++{ $as_echo "$as_me:${as_lineno-$LINENO}: result: $libgcc_cv_copysignl_lib" >&5
++$as_echo "$libgcc_cv_copysignl_lib" >&6; }
++
++case /${libgcc_cv_copysignl_lib}/ in
++  /-lm/) LIBGCC_LIBM="$LIBGCC_LIBM -lm" ;;
++  *) LIBGCC_LIBM= ;;
++esac
++
+ 
+ # Conditionalize the makefile for this target machine.
+ tmake_file_=
